detectors/azure/azurevm/vm.go

```diff
@@ -0,0 +1,116 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package azurevm // import "go.opentelemetry.io/contrib/detectors/azure/azurevm"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+)
+
+const (
+	defaultAzureVMMetadataEndpoint = "http://169.254.169.254/metadata/instance/compute?api-version=2021-12-13&format=json"
+)
+
+// AzureVMResourceDetector collects resource information of Azure VMs.
+type ResourceDetector struct {
+	endpoint string
+}
+
+type vmMetadata struct {
+	VMId       *string `json:"vmId"`
+	Location   *string `json:"location"`
+	ResourceId *string `json:"resourceId"`
+	Name       *string `json:"name"`
+	VMSize     *string `json:"vmSize"`
+	OsType     *string `json:"osType"`
+	Version    *string `json:"version"`
+}
+
+// New returns a [resource.Detector] that will detect Azure VM resources.
+func New() *ResourceDetector {
+	return &ResourceDetector{defaultAzureVMMetadataEndpoint}
+}
+
+// Detect detects associated resources when running on an Azure VM.
+func (detector *ResourceDetector) Detect(ctx context.Context) (*resource.Resource, error) {
+	jsonMetadata, runningInAzure, err := detector.getJSONMetadata()
```

Comment:
nit: not sure if it matters, but I would expect the context from Detect to be used with the http request.